services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java

```diff
@@ -592,18 +596,38 @@ private void initializeGenesis(final BeaconState genesisState) {
   private void onStoreInitialized() {
     UInt64 genesisTime = recentChainData.getGenesisTime();
     UInt64 currentTime = timeProvider.getTimeInSeconds();
-    UInt64 currentSlot = ZERO;
+    final UInt64 currentSlot;
     if (currentTime.compareTo(genesisTime) >= 0) {
       UInt64 deltaTime = currentTime.minus(genesisTime);
       currentSlot = deltaTime.dividedBy(SECONDS_PER_SLOT);
+      // Validate that we're running within the weak subjectivity period
+      validateLatestCheckpointIsWithinWeakSubjectivityPeriod(currentSlot);
     } else {
+      currentSlot = ZERO;
       UInt64 timeUntilGenesis = genesisTime.minus(currentTime);
       genesisTimeTracker = currentTime;
       STATUS_LOG.timeUntilGenesis(timeUntilGenesis.longValue(), p2pNetwork.getPeerCount());
     }
     slotProcessor.setCurrentSlot(currentSlot);
   }
 
+  private void validateLatestCheckpointIsWithinWeakSubjectivityPeriod(final UInt64 currentSlot) {
+    SafeFuture.of(() -> recentChainData.getStore().retrieveFinalizedCheckpointAndState())
+        .thenAccept(
+            finalizedCheckpointState -> {
+              final UInt64 slot = currentSlot.max(recentChainData.getCurrentSlot().orElse(ZERO));
+              weakSubjectivityValidator.validateLatestFinalizedCheckpoint(
+                  finalizedCheckpointState, slot);
+            })
+        .exceptionally(
+            err -> {
+              weakSubjectivityValidator.handleValidationFailure(
+                  "Encountered an error while trying to validate latest finalized checkpoint", err);
+              throw new RuntimeException(err);
+            })
+        .reportExceptions();
```

Comment:
This feels like a weird construct to both handle the error and pass it to the `weakSubjectivityValidator` and also leave it wind up handled by the uncaught exception handler.

I'd have though just reporting it to the validator would be enough and it can decide how loudly to report it. So something like:
```suggestion
        .finish(
            err -> {
              weakSubjectivityValidator.handleValidationFailure(
                  "Encountered an error while trying to validate latest finalized checkpoint", err);
            });
```

Reply:
Yeah - that seems like a better approach 👍